File: app/src/main/java/com/fourkplus/tvplayer/ui/design/Backdrop.kt
package com.fourkplus.tvplayer.ui.design

import androidx.compose.animation.core.Animatable
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshotFlow
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import com.fourkplus.tvplayer.BuildConfig
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import coil.compose.AsyncImagePainter
import coil.compose.rememberAsyncImagePainter
import coil.imageLoader
import coil.request.ImageRequest
import coil.size.Precision
import coil.size.Scale
import com.fourkplus.tvplayer.R
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.first

/**
 * What the app is currently asking the cinematic background to show.
 *
 * Only movies and series drive the background. Channels, Search, Settings and Language deliberately
 * leave it alone - they call nothing, and whatever artwork was last shown stays up. That is what
 * makes moving through the app feel like one continuous room rather than a slideshow.
 */
@Stable
class BackdropState {
    /** The artwork the most recently focused eligible item wants shown. */
    internal var requestedModel by mutableStateOf<String?>(null)
        private set

    /**
     * Whether the background follows what the viewer is looking at.
     *
     * False is Classic: the app's own artwork stays up everywhere and nothing replaces it. Some
     * people want the room to hold still - a picture that changes every time the highlight moves
     * is the single busiest thing on screen, and on a television that is across the room from you
     * it is movement in the corner of your eye all evening.
     *
     * Gated here rather than at the callers. Every page asks for artwork in the ordinary way and
     * this decides whether the request is honoured, so Classic cannot be defeated by a screen that
     * forgot to check.
     */
    var followsFocus by mutableStateOf(true)

    /**
     * Ask for [model] (a URL) to become the background. A blank or null [model] is a no-op: an item
     * with no artwork keeps whatever is already there instead of dropping the viewer onto the
     * default image, which would read as a flash every time an untagged title is focused.
     */
    fun show(key: String?, model: String?) {
        if (!followsFocus) return
        if (model.isNullOrBlank() || key.isNullOrBlank()) return
        if (model == requestedModel) return
        requestedModel = model
    }

    /**
     * Asks for the app's own default artwork back, fading whatever is up back out to it.
     *
     * Home uses this: it is where the app starts over, so it shows the app's own picture rather
     * than whichever title the viewer was last looking at somewhere else. A reset immediately
     * followed by a real request is collapsed by the same debounce every other change goes through,
     * so arriving somewhere that puts a picture up straight away never flashes the default first.
     */
    fun reset() {
        requestedModel = null
    }
}

@Composable
fun rememberBackdropState(): BackdropState = remember { BackdropState() }

/**
 * The app-wide background: the default artwork at the bottom, the settled image over it, and the
 * incoming image fading in over both.
 *
 * The settled layer is a [Painter] kept from the last image that finished loading - not a second
 * request for the same URL. Nothing ever asks Coil to re-fetch what is already on screen, so the
 * visible image cannot blink while a new one arrives. The incoming layer only becomes visible once
 * Coil reports it fully decoded, and a failed load simply never fades in, leaving the previous
 * artwork exactly where it was.
 *
 * The handover at the end of a fade is ordered so that every intermediate frame shows the new
 * image, from one layer or the other or both: the settled painter is replaced first, and only then
 * is the incoming layer released.
 */
@OptIn(FlowPreview::class)
@Composable
fun CinematicBackdrop(
    state: BackdropState,
    modifier: Modifier = Modifier,
    scrim: Brush = Tone.pageScrim(),
    content: @Composable BoxScope.() -> Unit
) {
    val reducedMotion = LocalReducedMotion.current

    var settled by remember { mutableStateOf<Painter?>(null) }
    var settledModel by remember { mutableStateOf<String?>(null) }
    var incomingModel by remember { mutableStateOf<String?>(null) }
    val fade = remember { Animatable(0f) }
    // Fades the settled layer back out when the app asks for its own artwork again, so returning to
    // the default looks like every other change of background rather than a cut.
    val settledFade = remember { Animatable(1f) }

    val context = LocalContext.current
    // Backdrops are decoded at up to 4K rather than at the size of the view they land in. Coil
    // sizes a request to its composable by default, which on a 1080p panel throws away every pixel
    // beyond 1920 wide before the image is ever drawn - so a 4K still and a 1080p one ended up
    // identical on screen, and identical again on a 4K panel where the difference would show.
    //
    // It never invents detail: this is a ceiling, not a target. A source smaller than 4K is decoded
    // whole, at its own resolution, because that is all there is to decode.
    val incomingRequest = remember(incomingModel) {
        incomingModel?.let {
            ImageRequest.Builder(context)
                .data(it)
                .size(BackdropWidthPx, BackdropHeightPx)
                .scale(Scale.FILL)
                .precision(Precision.INEXACT)
                .build()
        }
    }
    val incomingPainter = rememberAsyncImagePainter(model = incomingRequest, contentScale = ContentScale.Crop)

    // Rapid D-pad movement produces a request per item. Debouncing means only the item the viewer
    // actually settles on costs an image load, while the ones they scrolled past cost nothing.
    LaunchedEffect(state) {
        snapshotFlow { state.requestedModel }
            .distinctUntilChanged()
            .debounce(Motion.BackdropDebounceMs)
            .collect { model ->
                // A null model is a request for the app's own artwork back - see [BackdropState.
                // reset]. The settled layer fades out and the default drawn underneath is what is
                // left. Note this arrives through the same debounce as everything else, so a reset
                // immediately followed by a real request never shows: the viewer goes straight to
                // the new picture instead of flashing through the default on the way.
                if (model == null) {
                    if (settled != null) {
                        incomingModel = null
                        if (reducedMotion) {
                            settledFade.snapTo(0f)
                        } else {
                            settledFade.animateTo(0f, Motion.backdrop())
                        }
                        settled = null
                        settledModel = null
                        settledFade.snapTo(1f)
                    }
                    return@collect
                }
                // Already showing it, or already loading it: starting the transition again would
                // fade the picture out and back into itself for no reason.
                if (model == settledModel || model == incomingModel) return@collect
                fade.snapTo(0f)
                incomingModel = model
            }
    }

    // One pass per requested image: wait for that image to resolve, fade it up, hand it over.
    // Keyed on the model alone, so the effect cannot be restarted by the handover it performs -
    // an earlier version keyed it on derived state and re-ran on its own completion, which made
    // the background oscillate between the last two images forever.
    LaunchedEffect(incomingModel) {
        val requested = incomingModel ?: return@LaunchedEffect
        val result = snapshotFlow { incomingPainter.state }
            .first { it is AsyncImagePainter.State.Success || it is AsyncImagePainter.State.Error }
        if (result !is AsyncImagePainter.State.Success) {
            // Not an error the viewer should see: the artwork simply never appears.
            incomingModel = null
            return@LaunchedEffect
        }
        if (reducedMotion) fade.snapTo(1f) else fade.animateTo(1f, Motion.backdrop())
        // Order matters. The settled layer takes the finished image first, so the frame in which
        // the incoming layer disappears is already showing that same image underneath.
        settled = result.painter
        settledModel = requested
        incomingModel = null
        fade.snapTo(0f)
    }

    Box(modifier.fillMaxSize().background(Brush.verticalGradient(listOf(Tone.PageTop, Tone.PageBottom)))) {
        // Always present underneath, so the screen is never empty even before any artwork loads.
        //
        // The artwork is 16:9, which is a television exactly and nothing like a phone held
        // sideways - a 20:9 handset is far wider for its height, so filling the width leaves the
        // picture taller than the screen and Crop takes the difference off both ends. Centred,
        // half of that came off the top, which is where the astronaut's helmet is. Aligning to the
        // top instead spends the whole overlap at the bottom, on empty ground, and brings the head
        // back down into view. The television is 16:9 itself, has no overlap to place, and keeps
        // the centring it has always had.
        Image(
            painter = painterResource(R.drawable.bg_app_default),
            contentDescription = null,
            contentScale = ContentScale.Crop,
            alignment = if (BuildConfig.TOUCH_BUILD) Alignment.TopCenter else Alignment.Center,
            modifier = Modifier.fillMaxSize()
        )
        settled?.let { current ->
            Image(
                current,
                null,
                Modifier.fillMaxSize().alpha(settledFade.value),
                contentScale = ContentScale.Crop
            )
        }
        if (incomingModel != null) {
            Image(
                incomingPainter,
                null,
                Modifier.fillMaxSize().alpha(fade.value),
                contentScale = ContentScale.Crop
            )
        }
        Box(Modifier.fillMaxSize().background(Tone.sideScrim()))
        Box(Modifier.fillMaxSize().background(scrim))
        content()
    }
}

/** The ceiling a backdrop is decoded at. Sources below it are decoded whole; nothing is upscaled. */
private const val BackdropWidthPx = 3840
private const val BackdropHeightPx = 2160

/**
 * Warms Coil's cache with artwork the viewer is about to reach, so a deliberate move along a row
 * shows its background immediately instead of after a round trip. Fire-and-forget: failures are
 * irrelevant, and the requests are plain cache fills with no UI attached.
 *
 * Sized to match the backdrop's own request, so the cached entry is the one the backdrop will ask
 * for. A preload at a different size fills the cache with a bitmap the backdrop then ignores.
 */
@Composable
fun PreloadBackdrops(urls: List<String>) {
    val context = LocalContext.current
    LaunchedEffect(urls) {
        urls.filter { it.isNotBlank() }.take(12).forEach { url ->
            runCatching {
                context.imageLoader.enqueue(
                    ImageRequest.Builder(context)
                        .data(url)
                        .size(BackdropWidthPx, BackdropHeightPx)
                        .scale(Scale.FILL)
                        .precision(Precision.INEXACT)
                        .build()
                )
            }
        }
    }
}

/** A flat scrim for screens that sit on top of the backdrop without their own artwork. */
val PanelScrim: Brush = Brush.verticalGradient(listOf(Color(0xD90A1120), Color(0xF2050A14)))

File: web/src/ui/backdrop.ts
/**
 * The app-wide background, ported from Backdrop.kt.
 *
 * Three layers: the app's own artwork underneath, the settled image over it, and the incoming one
 * fading in over both. The handover is ordered so every frame shows something - the settled layer
 * takes the finished image first, and only then is the incoming layer released. Done the other
 * way round, the background blinks through to the default between the two.
 *
 * An image that fails to load simply never fades in, leaving what was there. A backdrop is
 * decoration; a broken one must not become a visible event.
 *
 * Rules carried over:
 *
 *  - Only films and series drive it. Channels, search and settings leave it alone, and whatever
 *    was last shown stays up - that is what makes moving through the app feel like one room
 *    rather than a slideshow.
 *  - Requests are debounced. Holding the D-pad down through twenty posters should cost one image
 *    request, not twenty.
 *  - Home shows the app's own artwork until the viewer actually moves onto something, so opening
 *    the app does not immediately replace it with a title nobody chose.
 */
import { Motion } from '../shared/tokens.generated';

/** How long the highlight has to rest on a title before its background is fetched. */
const WARM_AFTER_MS = 300;

export class Backdrop {
  private readonly base: HTMLElement;
  private readonly settled: HTMLImageElement;
  private readonly incoming: HTMLImageElement;
  private settledUrl: string | null = null;
  private incomingUrl: string | null = null;
  private timer: number | null = null;
  private warmTimer: number | null = null;
  /**
   * Pictures fetched and decoded ahead of being shown, newest last. Held so the engine keeps them:
   * an Image nobody references is free to be dropped, decoded pixels and all.
   */
  private readonly warm = new Map<string, HTMLImageElement>();

  constructor(host: HTMLElement) {
    this.base = document.createElement('div');
    this.base.className = 'backdrop-base';
    // Relative to the page, so it resolves the same while developing and inside the widget.
    this.base.style.backgroundImage = "url('./bg_app_default.png')";
    this.settled = document.createElement('img');
    this.settled.className = 'backdrop-layer';
    this.incoming = document.createElement('img');
    this.incoming.className = 'backdrop-layer';
    const scrim = document.createElement('div');
    scrim.className = 'backdrop-scrim';
    host.append(this.base, this.settled, this.incoming, scrim);
  }

  /**
   * Whether the background follows what the viewer is looking at.
   *
   * False is Classic: the app's own artwork stays up everywhere and nothing replaces it. Some
   * people want the room to hold still - a picture that changes every time the highlight moves is
   * the busiest thing on screen, and on a television across a room it is movement in the corner of
   * the eye all evening.
   *
   * Gated here rather than at the callers, exactly as on Android. Every page asks for artwork in
   * the ordinary way and this decides whether the request is honoured, so Classic cannot be
   * defeated by a screen that forgot to check.
   */
  followsFocus = true;

  /**
   * Ask for [url] to become the background. A blank url is a no-op rather than a reset: a title
   * with no artwork keeps whatever is already there, instead of dropping the viewer onto the
   * default image, which reads as a flash every time an untagged title is focused.
   */
  show(url: string | null | undefined): void {
    if (!this.followsFocus) return;
    if (!url) return;
    if (url === this.settledUrl || url === this.incomingUrl) return;
    this.schedule(url);
  }

  /** Back to the app's own artwork, fading rather than cutting. */
  reset(): void {
    this.schedule(null);
  }

  /**
   * Fetches and decodes [urls] now, so that when one of them becomes the background it is already
   * there. PreloadBackdrops on Android, with its limit of twelve.
   *
   * On Android the three-second wait before the background changes is spent with the picture
   * already warm - the neighbours of the focused card are fetched and decoded ahead of time - so
   * when the wait ends the picture is simply shown. Here the wait ended and *then* the download
   * began, followed by a decode of a full-size still on a television's processor, which is the
   * slowness the owner compared against the Android build.
   */
  preload(urls: (string | null | undefined)[]): void {
    if (!this.followsFocus) return;
    for (const url of urls.filter((value): value is string => Boolean(value)).slice(0, 12)) {
      const known = this.warm.get(url);
      if (known) {
        // Most recently wanted goes to the back, so trimming drops what is furthest behind.
        this.warm.delete(url);
        this.warm.set(url, known);
        continue;
      }
      const image = new Image();
      image.decoding = 'async';
      image.onload = () => {
        if (typeof image.decode === 'function') image.decode().catch(() => undefined);
      };
      image.src = url;
      this.warm.set(url, image);
    }
    // Twice the batch, so the pictures just passed are still warm when the viewer steps back.
    while (this.warm.size > 24) this.warm.delete(this.warm.keys().next().value!);
  }

  private schedule(url: string | null): void {
    if (this.timer !== null) window.clearTimeout(this.timer);
    if (this.warmTimer !== null) window.clearTimeout(this.warmTimer);
    this.warmTimer = null;
    this.timer = window.setTimeout(() => this.apply(url), Motion.BackdropDebounceMs);
    /*
     * The picture itself is asked for much sooner than it is shown.
     *
     * The three seconds are Android's, and they are about when the background *changes* - for
     * somebody who has stopped to look, not somebody passing through. They were never meant to
     * be three seconds and then a download. A short dwell first still keeps a held key from
     * fetching every title it passes; after that the fetch runs during the wait instead of after.
     */
    if (url) this.warmTimer = window.setTimeout(() => this.preload([url]), WARM_AFTER_MS);
  }

  private apply(url: string | null): void {
    if (url === null) {
      this.settled.style.opacity = '0';
      this.incoming.style.opacity = '0';
      this.settledUrl = null;
      this.incomingUrl = null;
      return;
    }
    this.incomingUrl = url;
    const image = new Image();
    image.decoding = 'async';
    image.onload = async () => {
      /*
       * Decoded before it is shown, not while it fades.
       *
       * A backdrop is a full-screen photograph, and an image that has loaded has not been decoded:
       * that happens on first paint, on the main thread, in the middle of the fade - a visible
       * hitch in the animation and a key press that waits behind it. decode() does it beforehand,
       * off the main thread where the engine can. It is decoded here, on a loose Image, and shown
       * on the layer: Chromium keeps one decoded copy per loaded image, which the layer showing the
       * same URL reuses. Not on every engine this app runs on, and a failure only means the old
       * behaviour.
       */
      if (typeof image.decode === 'function') await image.decode().catch(() => undefined);
      // Overtaken while this was decoding - a faster neighbour won. Drop it rather than fading in
      // a picture the viewer has already moved past.
      if (this.incomingUrl !== url) return;
      this.incoming.src = url;
      this.incoming.style.opacity = '1';
      window.setTimeout(() => {
        if (this.incomingUrl !== url) return;
        // Settled first, then release the incoming layer: both show the same picture for a frame,
        // which is what stops the swap being visible.
        this.settled.src = url;
        this.settled.style.opacity = '1';
        this.settledUrl = url;
        this.incoming.style.opacity = '0';
        this.incomingUrl = null;
      }, Motion.BackdropMs);
    };
    image.onerror = () => {
      if (this.incomingUrl === url) this.incomingUrl = null;
    };
    image.src = url;
  }
}
